third_party/amd/lib/TritonAMDGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp

```diff
@@ -193,18 +193,82 @@ bool isColMajor(::llvm::ArrayRef<unsigned> order) {
   return order[0] == (rank - 2);
 }
 
+/// Expand layout of dot operands and dot results to 3d variant.
+///
+/// If given layout describes 3d tensor, return it without change.
+/// If given layout describes 2d tensor, create new layout
+/// describing 3d tensor by adding batch = 1.
+Attribute getExpandedEncoding(Attribute encoding) {
+  auto ctx = encoding.getContext();
+  if (auto sharedEncoding = mlir::dyn_cast<SharedEncodingAttr>(encoding)) {
+    auto expandedEncoding = getExpandedSharedEncoding(sharedEncoding);
+    return expandedEncoding;
+  } else if (auto mfmaEncoding =
+                 mlir::dyn_cast<AMDMfmaEncodingAttr>(encoding)) {
+    auto warpsPerCTA = triton::gpu::getWarpsPerCTA(mfmaEncoding);
+    auto rank = warpsPerCTA.size();
+    if (rank == 3) {
+      return encoding;
+    }
+ 
+    SmallVector<unsigned, 3> expandedWarpsPerCTA{1, warpsPerCTA[0], warpsPerCTA[1]};
+    auto expandedMfmaEncoding = AMDMfmaEncodingAttr::get(
+        ctx, mfmaEncoding.getVersionMajor(), mfmaEncoding.getVersionMinor(), expandedWarpsPerCTA,
+        mfmaEncoding.getMDim(), mfmaEncoding.getNDim(), mfmaEncoding.getIsTransposed(),
+        getExpandedCTALayout(ctx, mfmaEncoding.getCTALayout()));
+
+    return expandedMfmaEncoding;
+  } else if (auto dotOperandEncoding =
+                 mlir::dyn_cast<DotOperandEncodingAttr>(encoding)) {
+    auto mfmaEncoding =
+        mlir::cast<AMDMfmaEncodingAttr>(dotOperandEncoding.getParent());
+    auto expandedMfmaEncoding = getExpandedEncoding(mfmaEncoding);
+    auto expandedEncoding = DotOperandEncodingAttr::get(
+        ctx, dotOperandEncoding.getOpIdx(), expandedMfmaEncoding,
+        dotOperandEncoding.getKWidth());
+    return expandedEncoding;
+  } else
+    llvm_unreachable("unsupported encoding");
+}
+
+/// Expand date type of dot operands to 3d variant. If the given type is a 3d tensor,
```

Comment:
nitpick :)
```suggestion
/// Expand data type of dot operands to 3d variant. If the given type is a 3d tensor,
```

or maybe even 
```suggestion
/// Expand type of dot operands to 3d variant. If the given type is a 3d tensor,
```
because we typically use "data type" to describe type of tensor elements, like `float32` or `int8`

I suggest to move this function in a common place as well, it is used for both mfma and wmma converters.